### ISSUE TITLE
Make Tweet Video Resolution Configurable

### ISSUE THREAD

**Issue by mackenzie-remote** [oysttyer/oysttyer#77]
Currently the smallest bitrate is chosen for video mp4 links. Could this be made configurable?

https://github.com/oysttyer/oysttyer/blob/master/oysttyer.pl#L7429-L7441

**Comment by atomicules** [comment]
It could be.

But I'm not going to do it just yet. I have something else I maintain 
that I need to spend some time on first and then the next focus for 
Oysttyer will be the code re-organisation as hopefully that would make 
it easier for others to contribute.

So I'll have to come back to this.

**Comment by trs80** [comment]
Also apparently .mp4 URLs are going away: https://twittercommunity.com/t/retiring-mp4-video-output-support-on-august-1st-2016/66045 but they're still available at the moment.

**Comment by atomicules** [comment]
Thanks for the heads up.

**Comment by atomicules** [comment]
Hmmm... I've just found a new tweet with an mp4 only (no m3u8).

Twitter!

**Comment by mackenzie-remote** [comment]
OP here! Now Twitter seemed to have abandoned their M3U8 experiment and reverted to MP4 would it be possible to reopen this issue?

**Comment by funnelfiasco** [comment]
Consider it reopened!

The difficulty here is that the bitrate isn't a label, but a number (which makes sense but is inconvenient for our purposes). I think what we'd do for now is allow the user to specify "lowest" or "highest" and then sort the variants based on that. If there are more than two variants, (which it looks like there can be), then you'll have to live with the two options.

It's worth noting, too, that sometimes the `bitrate` can be 0 (see [this tweet](https://twitter.com/RyanMaue/statuses/872091091994107905), for example).

I'd planned on pushing 2.8.2 out later this week. If I can get to this in the next few days, I'll delay that to include this, otherwise I'll push it off to the next release.

Note to self: see [this section](https://dev.twitter.com/overview/api/entities-in-twitter-objects#extended-entities-json-example-for-video) of the API docs for an example.

**Comment by funnelfiasco** [comment]
@mackenzie-remote I bet you thought I forgot about you! Here's a branch that implements this feature. The default is to set `video_bitrate=largest` and if `video_bitrate` is anything but `largest` it will give you the smallest. We may be able to improve on this later, but for now I figure most people will want one or the other.

https://github.com/oysttyer/oysttyer/blob/feature/i77-configurable_video/oysttyer.pl

I'll give this a try for a few days and if you're happy with it, too, I'll merge it in for 2.9, which will come out realsoonnow.

**Comment by mackenzie-remote** [comment]
Excellent, thanks @funnelfiasco !

**Comment by funnelfiasco** [comment]
Gah! There's a definite bug here. Looks like GIFs are being replaced with the last-seen video? Not entirely sure yet, but it's definitely in need of more work.

**Comment by funnelfiasco** [comment]
Okay, the branch is updated with a fix. Always initialize your arrays, kids.

**Comment by mackenzie-remote** [comment]
@funnelfiasco I've updated to the array fix - unfortunately I'm seeing issues where the video/gif link is from the wrong tweet.

EDIT: If I grep my log I can see the same mp4 URL is attached to multiple tweets.

**Comment by funnelfiasco** [comment]
Hm, I thought my fix...fixed that. Can you give me any more information?
For example do you get a tweet from one that appeared fairly recently? A
couple of status IDs might help.

**Comment by mackenzie-remote** [comment]
This is one tweet I've been testing - https://twitter.com/MalwareJake/status/876140051880701952 - each time I fire up oysttyer on this branch it comes up with a different mp4 (from other tweets in my timeline).

**Comment by funnelfiasco** [comment]
Thanks! I believe what was happening here is that my array initialization fixed part of the issue, but not all of it. I had also thought to initialize the hashes used to find the right bitrate, but I implemented that incorrectly.

The first fix made everything except GIFs work (or at least mostly work), but GIFs have a `bitrate` of 0. I've pushed a fixed fix which seems to actually work. At least, I'm no longer seeing the behavior @mackenzie-remote reported.
